### DISCUSSION TITLE
Id=0 and nan values in DiagTrackParticles (I think it is not physical problem)

### DISCUSSION THREAD

**Discussion by akbwyfc** [SmileiPIC/Smilei#729] [Q&A] [answered]
Hi All,

I'm trying to trace particles in a 3D simulation. The results seems Physical, but when I trace, I figured out that I have a lot of nan values during the DiagTrackParticles in the places where Id=0 as follows:

Id at each 10 time steps: [7177 7177    0    0 7177    0    0    0    0    0    0 7177    0    0]
 z at each 10 time steps[23.91445166 23.93263856         nan         nan 23.98661158         nan nan         nan         nan         nan         nan 24.10842363 nan         nan]

(it is the same for y,z,pxyz as well)
As you can see when id is zero, the z value is nan, otherwise it is correct. 
Also, I checked the 3D magnetic field values and I do not have nan value in B field.
![Uelm](https://github.com/SmileiPIC/Smilei/assets/55910280/00592d7d-918c-4788-839b-e70ee20eebea)

Also, following please find the magnetic field energy of the simulation.
The box size is 128x128x128 with 8 particles per cell and the boundaries are periodic.
I use 32 cores of openmp and i set the patch size as: number_of_patches = [4,4,4]

Thanks in advance for your help.
Cheers

**Comment by mccoys** [comment]
These results usually correspond to particles that leave the box, or are filtered out (temporarily) by the filter function

**Comment by akbwyfc** [comment]
Thanks @mccoys for your reply, I think as the boundaries are periodic, particles will not leave the box, 
Also, regarding filtered particles, if I set filter as: 5<pz<6 for electrons, if pz would be >6 in later time steps, does it send 0 id and 
and again later if 5<pz<6 would it return the correct id?

In this case, how can I trace a particle energy that for example initially 5<pz<6.

Cheers

**Comment by mccoys** [comment]
Yes the times where the particle is out will be filled (by happi post-processing) with either 0 or NaN.

If you want to filter particles at the initial timestep, and keep track of those at later times, I suggest you read https://smileipic.github.io/Smilei/Use/ids.html
In the filter function you can use `if Main.iteration == ...` to choose when your condition is taken, and in this case, set one of the available bits to 1. For example by setting `particles.id += 2**63`. Then, for the following timesteps, simply track particles that have an ID larger than `2**63`

**Reply by akbwyfc** [reply]
Thanks @mccoys for your help, I will try that. BTW, one more question to be sure, for example I want to trace electrons with Ekin=100-110Mev and as I can use only gamma.v in the DiagTrackParticles, is it exact if I filter 100.995<P<110.995:
(gamma-1)m_ec^2=100 --> gamma=101 -->v**2=1-(1/gamma)**2-->V=0.999951, gamma.v=100.995

Cheers

**Reply by mccoys** [reply]
You don't have v in the filter function, but p=gamma*v, which is also sqrt(gamma^2-1). Note that i am using the normalization in smilei. Using this you should be able to calculate the energy

**Comment by akbwyfc** [comment]
Thanks @mccoys for your explanations, I tried to set filter as:
def filt(p):
    import numpy
    if Main.iteration == 10:
  

      if p.pz> 300.:
          p.id += 40000000
      return false
      else:
       return p.id > 40000000

but I receive following error:

 On rank 0 [Python] ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
 --------------------------------------------------------------------------------
 [ERROR](0) src/Diagnostic/DiagnosticParticleList.cpp:195 (run) A DiagTrackParticles filter has not provided a correct result

I would be thankful for your help.

Cheers

**Comment by mccoys** [comment, marked as answer]
Several issues here
* False has a capital letter in python
* You should not `return False` as smilei requires that you return an array of booleans of the same size as the particle array. Instead do something like `p.id[p.pz>300] += 2**63`
* I told you above to use one of the last 8 bits. For instance with 2**63. You might need to learn about integer representation

**Comment by akbwyfc** [comment]
Thanks @mccoys for your reply and the detailed explanations.
I tried and I think it is OK now, Yes, I will try to learn more about integer representation based on bits.

Cheers
